File: py/deepgene/mutant_extractor.py
# ABOUTME: Module for extracting mutation mentions from biomedical text
# ABOUTME: Uses DSPy with Gemini to identify genetic variants, SNPs, and mutations

import logging
import dspy

logger = logging.getLogger(__name__)


class MutantExtraction(dspy.Signature):
    """Extract mutation and variant mentions from biomedical text.

    Identify all genetic mutations, variants, and SNPs mentioned in the text.
    Include rs numbers, protein mutations (p. notation), DNA mutations (c. notation),
    and simple mutation names (e.g., V600E).
    """

    text: str = dspy.InputField(desc="Abstract or paper text to analyze for mutations")
    mutants: list[str] = dspy.OutputField(
        desc="List of all mutation identifiers found: rs numbers (rs116515942), "
             "protein mutations (p.Gly12Asp), DNA mutations (c.35G>A), "
             "simple mutations (V600E), gene variants, SNPs"
    )


class MutantExtractor:
    """Extract mutation mentions from biomedical text using DSPy/Gemini."""

    def __init__(self):
        self.extractor = dspy.ChainOfThought(MutantExtraction)

    def extract_mutants(self, text: str) -> list[str]:
        """
        Extract mutation mentions from text.

        Args:
            text: Abstract or paper text to analyze

        Returns:
            List of mutant identifiers like:
            - "rs116515942"
            - "p.Gly12Asp"
            - "c.35G>A"
            - "V600E"
        """
        if not text or len(text) < 10:
            return []

        try:
            text_sample = text[:2000]

            result = self.extractor(text=text_sample)

            if result.mutants and isinstance(result.mutants, list):
                logger.info(f"Extracted {len(result.mutants)} mutants from text")
                return result.mutants

            return []

        except Exception as e:
            logger.error(f"NER extraction failed: {e}")
            return []


_extractor_instance = None


def get_mutant_extractor() -> MutantExtractor:
    """Get cached extractor instance (reuses DSPy configuration)."""
    global _extractor_instance
    if _extractor_instance is None:
        logger.info("Initializing mutant extractor")
        _extractor_instance = MutantExtractor()
    return _extractor_instance

File: js/src/lib/mutant_extractor.ts
// ABOUTME: Module for extracting mutation mentions from biomedical text
// ABOUTME: Uses Vercel AI SDK (model-agnostic) with Gemini to identify genetic variants, SNPs, and mutations

import { generateObject } from 'ai';
import { createGoogleGenerativeAI } from '@ai-sdk/google';
import { z } from 'zod';

let provider: ReturnType<typeof createGoogleGenerativeAI> | null = null;

/**
 * Initialize with API key.
 */
export function initializeAI(key: string): void {
  provider = createGoogleGenerativeAI({ apiKey: key });
}

/**
 * Mutant extraction schema (like DSPy signature).
 */
const MutantExtractionSchema = z.object({
  mutants: z
    .array(z.string())
    .describe(
      'List of all mutation identifiers found: rs numbers (rs116515942), protein mutations (p.Gly12Asp), DNA mutations (c.35G>A), simple mutations (V600E), gene variants, SNPs'
    ),
});

/**
 * Extract mutation mentions from biomedical text.
 *
 * Returns list of mutant identifiers like:
 * - "rs116515942"
 * - "p.Gly12Asp"
 * - "c.35G>A"
 * - "V600E"
 */
export async function extractMutants(text: string): Promise<string[]> {
  if (!text || text.length < 10) {
    return [];
  }

  if (!provider) {
    console.error('Provider not initialized. Call initializeAI() first.');
    return [];
  }

  try {
    const textSample = text.slice(0, 2000);

    const result = await generateObject({
      model: provider('gemini-2.0-flash-exp'),
      schema: MutantExtractionSchema,
      prompt: `Extract all mutation and variant mentions from the following biomedical text.

Identify genetic mutations, variants, and SNPs mentioned in the text.
Include rs numbers (rs116515942), protein mutations (p.Gly12Asp), DNA mutations (c.35G>A), and simple mutation names (V600E).

Text:
${textSample}`,
    });

    const mutants = result.object.mutants;

    if (Array.isArray(mutants) && mutants.length > 0) {
      console.log(`Extracted ${mutants.length} mutants from text`);
      return mutants;
    }

    return [];
  } catch (error) {
    console.error('NER extraction failed:', error);
    return [];
  }
}

let extractorInitialized = false;

/**
 * Get extractor instance (ensures AI is initialized).
 */
export function getExtractor(key: string): { extractMutants: typeof extractMutants } {
  if (!extractorInitialized) {
    console.log('Initializing mutant extractor');
    initializeAI(key);
    extractorInitialized = true;
  }
  return { extractMutants };
}
